### PR TITLE
[Enhance] Support training on IPU and add fine-tuning configs of ViT.

### PR DESCRIPTION
**Pull request by HuDi2018**
need use a [PR](https://github.com/open-mmlab/mmcv/pull/1769) in mmcv
see and run tools/ipu_train_example.sh for details.

## Motivation

run mmclassification ViT training on IPU

## Modification

add some arguments parse in train.py and test.py to make a preparation for IPU training/inference

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

bash tools/ipu_train_example.sh

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by Ezra-Yu** [comment]
please make sure that all the commitors have signed the CLA. If the user is untracked, you can use the git-rebase to modify the git message

**Comment by HumberMe** [comment]
> untracked

I will rebase commits and re-submit a PR

File: .gitignore [hunk 1]
```diff
@@ -123,3 +123,12 @@ venv.bak/
...
 *.pth
+
+# IPU
+*.pvti
+*.pvti-journal
+/cache_engine
+/report
+
+# my dev
+/configs/myconfigs
```
Comment (mzr1996) [inline .gitignore:134]: Please remove the private ignore path.

File: configs/vision_transformer/vit-base-p16_ft-4xb544_in1k-224_ipu.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
+_base_ = [
+    '../_base_/models/vit-base-p16.py',
+    '../_base_/datasets/imagenet_bs64_pil_resize_autoaug.py',
+    '../_base_/schedules/imagenet_bs4096_AdamW.py',
```
Comment (mzr1996) [inline configs/vision_transformer/vit-base-p16_ft-4xb544_in1k-224_ipu.py:4]: Looks almost all settings in the `_base_/schedules/imagenet_bs4096_AdamW.py` is overrided, just remove the base config.

File: hdDebug.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import time
```
Comment (mzr1996) [inline hdDebug.py:1]: What's this file for?

File: configs/vision_transformer/vit-base-p16_ft-4xb544_in1k-224_ipu.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    drop_last=True,
+    train=dict(pipeline=train_pipeline),
+    val=dict(
+        pipeline=test_pipeline,
+        samples_per_gpu=4,
+        workers_per_gpu=1,),
+    test=dict(
+        pipeline=test_pipeline,
+        samples_per_gpu=4,
+        workers_per_gpu=1),)
```
Comment (mzr1996) [inline configs/vision_transformer/vit-base-p16_ft-4xb544_in1k-224_ipu.py:60]: Are these `samples_per_gpu` and `workers_per_gpu` arguments necessary?

File: configs/vision_transformer/vit-base-p16_ft-4xb544_in1k-224_ipu.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+train_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(type='RandomResizedCrop', size=224, backend='pillow'),
+    dict(type='RandomFlip', flip_prob=0.5, direction='horizontal'),
+    dict(type='Normalize', **img_norm_cfg),
+    dict(type='ImageToTensor', keys=['img']),
+    dict(type='ToTensor', keys=['gt_label']),
+    dict(type='Collect', keys=['img', 'gt_label']),
+    dict(type='ToHalf', keys=['img'])
```
Comment (mzr1996) [inline configs/vision_transformer/vit-base-p16_ft-4xb544_in1k-224_ipu.py:34]: please place the `ToHalf` transform after `ToTensor` and before `Collect`.
Usually, the `Collect` transform is the last transform in a pipeline.

File: mmcls/models/losses/accuracy.py [hunk 6]
```diff
@@ -56,6 +56,7 @@ def accuracy_torch(pred, target, topk=(1, ), thrs=0.):
     res = []
     maxk = max(topk)
     num = pred.size(0)
+    pred = pred.float()
```
Comment (mzr1996) [inline mmcls/models/losses/accuracy.py:59]: Is this line necessary?

File: tools/ipu_train_example.sh [hunk 7]
```diff
@@ -0,0 +1,9 @@
+
```
Comment (mzr1996) [inline tools/ipu_train_example.sh:1]: It's better to move this script to the `demo` folder, because this is an example instead of a tool.

**Review by mzr1996 — COMMENTED** [review]
Please rebase to the latest `dev` branch.

File: configs/vision_transformer/vit-base-p16_ft-4xb510_in1k-224_ipu_half.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    dict(type='Resize', size=(224, -1), backend='pillow'),
+    dict(type='CenterCrop', crop_size=224),
+    dict(type='Normalize', **img_norm_cfg),
+    dict(type='ImageToTensor', keys=['img']),
+    dict(type='Collect', keys=['img'])
+]
+
+# change batch size
+data = dict(
+    samples_per_gpu=510,
```
[the remark this answers is not in the document]
Comment (HumberMe) [reply configs/vision_transformer/vit-base-p16_ft-4xb510_in1k-224_ipu_half.py]: memory is not enough

File: hdDebug.py (hunk 3 above)
Comment (HumberMe) [reply hdDebug.py:1]: sorry, used for debug, I will delete it

File: configs/vision_transformer/vit-base-p16_ft-4xb544_in1k-224_ipu.py (hunk 4 above)
Comment (HumberMe) [reply configs/vision_transformer/vit-base-p16_ft-4xb544_in1k-224_ipu.py:60]: the partition way of train and inference are different(this is optimized model training config).
so the samples_per_gpu are different between train and val.

File: mmcls/models/losses/accuracy.py (hunk 6 above)
Comment (HumberMe) [reply mmcls/models/losses/accuracy.py:59]: when train in fp16, pred will be half, and torch cpu not implement topk on half

**Review by mzr1996 — APPROVED** [review]
LGTM
